keydata.go

```diff
@@ -482,110 +504,180 @@ func (d *KeyData) snapModelAuthKey(auxKey PrimaryKey) ([]byte, error) {
+	}
+	if params.EncryptionKeySize < 0 || params.EncryptionKeySize > 32 {
+		// The key size can't be larger than 32 with the supported cipher
+		return nil, nil, nil, fmt.Errorf("invalid encryption key size (%d bytes)", params.EncryptionKeySize)
+	}
+	if params.AuthKeySize < 0 {
+		return nil, nil, nil, fmt.Errorf("invalid auth key size (%d bytes)", params.AuthKeySize)
 	}
 
-	// Derive both a key and an IV from the passphrase in a single pass.
-	keyLen := passphraseEncryptionKeyLen + aes.BlockSize
+	kdfAlg := d.data.KDFAlg
+	if !kdfAlg.Available() {
+		return nil, nil, nil, fmt.Errorf("unavailable leaf KDF digest algorithm %v", kdfAlg)
+	}
+
+	// Include derivation parameters in the Argon2 salt in order to protect them
+	builder := cryptobyte.NewBuilder(nil)
+	builder.AddASN1(cryptobyte_asn1.SEQUENCE, func(b *cryptobyte.Builder) { // SEQUENCE {
+		b.AddASN1OctetString(params.KDF.Salt)                               // salt OCTET STRING
+		kdfAlg.marshalASN1(b)                                               // kdfAlgorithm AlgorithmIdentifier
+		b.AddASN1(cryptobyte_asn1.UTF8String, func(b *cryptobyte.Builder) { // encryption UTF8String
+			b.AddBytes([]byte(params.Encryption))
+		})
+		b.AddASN1Int64(int64(params.EncryptionKeySize)) // encryptionKeySize INTEGER
+		b.AddASN1Int64(int64(params.AuthKeySize))       // authKeySize INTEGER
+	})
+	salt, err := builder.Bytes()
+	if err != nil {
+		return nil, nil, nil, xerrors.Errorf("cannot serialize salt: %w", err)
+	}
 
-	params, err := kdfOptions.deriveCostParams(keyLen, kdf)
+	costParams := &KDFCostParams{
+		Time:      uint32(params.KDF.Time),
+		MemoryKiB: uint32(params.KDF.Memory),
+		Threads:   uint8(params.KDF.CPUs)}
+	derived, err := kdf.Derive(passphrase, salt, costParams, uint32(params.DerivedKeySize))
 	if err != nil {
-		return xerrors.Errorf("cannot derive KDF cost parameters: %w", err)
+		return nil, nil, nil, xerrors.Errorf("cannot derive key from passphrase: %w", err)
+	}
+	if len(derived) != params.DerivedKeySize {
+		return nil, nil, nil, errors.New("KDF returned unexpected key length")
 	}
 
-	var salt [16]byte
-	if _, err := rand.Read(salt[:]); err != nil {
-		return xerrors.Errorf("cannot read salt for new passphrase: %w", err)
+	key = make([]byte, params.EncryptionKeySize)
+	r := hkdf.Expand(func() hash.Hash { return kdfAlg.New() }, derived, []byte("PASSPHRASE-ENC"))
+	if _, err := io.ReadFull(r, key); err != nil {
+		return nil, nil, nil, xerrors.Errorf("cannot derive encryption key: %w", err)
+	}
+
+	iv = make([]byte, aes.BlockSize)
+	r = hkdf.Expand(func() hash.Hash { return kdfAlg.New() }, derived, []byte("PASSPHRASE-IV"))
+	if _, err := io.ReadFull(r, iv); err != nil {
+		return nil, nil, nil, xerrors.Errorf("cannot derive IV: %w", err)
 	}
 
-	key, err := kdf.Derive(passphrase, salt[:], params, uint32(keyLen))
+	auth = make([]byte, params.AuthKeySize)
+	r = hkdf.Expand(func() hash.Hash { return kdfAlg.New() }, derived, []byte("PASSPHRASE-AUTH"))
```

Comment:
if I change the constant  string no tests seem to fail, I wonder if and how this bit of the code is tested

Reply:
I added a test case for that for good measure. The test hardcodes a keydata generated with the current valid info fields.

Reply:
The added test looks ok - one note for future reference is that we have `testutil.DecodeHexString` for embedding bytes into tests which will fail the test if an invalid string is supplied without having to check an error.